rfcs/experimental/dynamic_selection/README.md

```diff
@@ -0,0 +1,571 @@
+4. Profiles each set of unique `f` and `args` for each resource in the default set and then uses the best.
+5. Uses the provided set of resources and profiles each set of unique `f` and `args` for each resource uses the best.
+
+#### select heuristic (expository)
+
+```cpp
+  template<typename Function, typename ...Args>
+  selection_type select(Function&& f, Args&&...args) {
+    if (initialized_) {
+      auto k = make_task_key(f, args...);
+      auto tuner = get_tuner(k);
+      auto offset = tuner->get_resource_to_profile();
+      if (offset == use_best) {
+        return selection_type {*this, tuner->best_resource_, tuner}; 
+      } else {
+        auto r = resources_[offset];
+        return selection{*this, r, tuner}; 
+      }
+    } else {
+      throw std::logic_error(“selected called before initialization”);
+    } 
+  }
+```
+
+#### execution info reporting requirements
+
+```cpp
+  void report(task_execution_time_t, timing_t v) { 
+    tuner_->add_timing(offset_, v); 
+  }
+```
+
+#### Exceptions
+
+Constructor or initialize may throw `std::bad_alloc` or `std::logic_error`.
+
+## Backends
+
+Backends allow generic policies to be implemented. End-users do not directly interact
+with a backend, except to choose the backend if they opt-out of the default SYCL backend.
+Custom policy writers that wish to implement a generic policy that can accept backends
+should follow the backend contract. Developers that wish to provide a backend that can
+be used with the existing concrete policies should also follow the backend contract.
+
+### Named Requirements
+
+The type `T` satisfies the *Backend* contract if given,
+
+- `b` an arbitrary identifier of type `T`
+- `args` an arbitrary parameter pack of types `typename… Args`
+- `s` is of type `S` and satisfies *Selection* and `is_same_v<resource_t<S>, resource_t<T>>` is true
+` `f` a function object with signature `wait_t<T> fun(resource_t<T>, Args…);`
+- `r` a container of resources of type `std::vector<resource_t<T>`.
+
+The following *must* be well-formed:
+
+`b.submit(s, f, args…)`
+
+Returns an object that satisfies *Submission*. The function
+invokes `f` but does not wait for the `wait_t<T>` object returned by it.
+
+`b.get_submission_group()`
+
+Returns an object that has a member function `void wait()`. Calling this
+wait function blocks until all previous submissions to this backend are
+complete.
+
+`b.get_resources()`
+
+Returns a `std::vector<resource_type>`.
+
+*optional: * `void lazy_report()`
+
+If defined by a backend, this function must be called before each new 
+selection. It triggers reporting of the necessary execution info back to
+the policy.
+
+## Free Functions
+
+`vector<typename policy_traits<P>::resource_type> get_resources(P&& p);`
+
+Returns the resources associated with the Policy `p`.
+
+May throw `std::bad_alloc` or `std::logic_error`. 
+
+A call to `get_resources` may cause lazy initialization. Initialization may throw `std::bad_alloc`.
```

Comment:
We probably need to define "lazy initialization" for a backend (and its resources).